### PR TITLE
hydra: support plugins

### PR DESCRIPTION
**Pull request by dberenbaum**
Closes https://github.com/iterative/dvc/issues/9731.

We weren't supporting Hydra plugins before, which are auto-discovered from `hydra_plugins`, because the path to `hydra_plugins` was not in `sys.path`. With this PR, we look for `hydra_plugins` at the root of the DVC repo unless a different path is configured with `dvc config hydra.plugins_path`.

TODO:
- Fix test issue
- Docs PR (also include docs for https://github.com/iterative/dvc/pull/9783)

### CODE REVIEW

**Comment by dberenbaum** [comment]
The test here pass for me when running the individual test but fail for me otherwise. I'm guessing I'm missing something about how pytest manipulates `sys.path` but not sure yet how to make the tests pass consistently.

Passing for individual test:

```sh
$ pytest tests/func/utils/test_hydra.py::test_compose_and_dump_plugins
================================== test session starts ==================================
platform darwin -- Python 3.11.7, pytest-7.4.4, pluggy-1.3.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/dave/Code/dvc
configfile: pyproject.toml
plugins: test-utils-0.0.8, timeout-2.2.0, flaky-3.7.0, cov-4.1.0, mock-3.12.0, docker-2.0.1, xdist-3.5.0, anyio-4.2.0, benchmark-4.0.0, lazy-fixture-0.6.3, hydra-core-1.3.2
collected 1 item

tests/func/utils/test_hydra.py .                                                  [100%]

=================================== 1 passed in 0.15s ===================================
```

Failing when run with other tests:

```sh
$ pytest tests/func/utils/test_hydra.py
================================== test session starts ==================================
platform darwin -- Python 3.11.7, pytest-7.4.4, pluggy-1.3.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/dave/Code/dvc
configfile: pyproject.toml
plugins: test-utils-0.0.8, timeout-2.2.0, flaky-3.7.0, cov-4.1.0, mock-3.12.0, docker-2.0.1, xdist-3.5.0, anyio-4.2.0, benchmark-4.0.0, lazy-fixture-0.6.3, hydra-core-1.3.2
collected 64 items

tests/func/utils/test_hydra.py .s..s..s..............s........................... [ 78%]
.....F........                                                                    [100%]

======================================= FAILURES ========================================
_____________________________ test_compose_and_dump_plugins _____________________________

tmp_dir = PosixTmpDir('/private/var/folders/24/99_tf1xj3vx8k1k_jkdmnhq00000gn/T/pytest-of-dave/pytest-41/test_compose_and_dump_plugins0')

    def test_compose_and_dump_plugins(tmp_dir):
        """Ensure Hydra plugins are loaded."""
        from dvc.utils.hydra import compose_and_dump

        config = tmp_dir / "conf" / "config.yaml"
        config.parent.mkdir()
        config.write_text("foo: '${plus_10:1}'\n")
        plugin_code = "from omegaconf import OmegaConf\n"
        plugin_code += "OmegaConf.register_new_resolver('plus_10', lambda x: x + 10)"
        plugin_file = tmp_dir / "hydra_plugins" / "resolver.py"
        plugin_file.parent.mkdir()
        plugin_file.write_text(plugin_code)
        output_file = tmp_dir / "params.yaml"
>       compose_and_dump(output_file, str(config.parent), None, "config", str(tmp_dir), [])

/Users/dave/Code/dvc/tests/func/utils/test_hydra.py:279:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/dave/Code/dvc/dvc/utils/hydra.py:63: in compose_and_dump
    OmegaConf.resolve(cfg)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/omegaconf.py:785: in resolve
    omegaconf._impl._resolve(cfg)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/_impl.py:40: in _resolve
    _resolve_container_value(cfg, k)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/_impl.py:14: in _resolve_container_value
    resolved = node._dereference_node()
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:246: in _dereference_node
    node = self._dereference_node_impl(throw_on_resolution_failure=True)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:277: in _dereference_node_impl
    return parent._resolve_interpolation_from_parse_tree(
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:584: in _resolve_interpolation_from_parse_tree
    resolved = self.resolve_parse_tree(
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:764: in resolve_parse_tree
    return visitor.visit(parse_tree)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/antlr4/tree/Tree.py:34: in visit
    return tree.accept(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar/gen/OmegaConfGrammarParser.py:206: in accept
    return visitor.visitConfigValue(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar_visitor.py:101: in visitConfigValue
    return self.visit(ctx.getChild(0))
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/antlr4/tree/Tree.py:34: in visit
    return tree.accept(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar/gen/OmegaConfGrammarParser.py:342: in accept
    return visitor.visitText(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar_visitor.py:298: in visitText
    return self.visitInterpolation(c)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar_visitor.py:125: in visitInterpolation
    return self.visit(ctx.getChild(0))
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/antlr4/tree/Tree.py:34: in visit
    return tree.accept(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar/gen/OmegaConfGrammarParser.py:1041: in accept
    return visitor.visitInterpolationResolver(self)
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/grammar_visitor.py:179: in visitInterpolationResolver
    return self.resolver_interpolation_callback(
/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:750: in resolver_interpolation_callback
    return self._evaluate_custom_resolver(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = {'foo': '${plus_10:1}'}, key = 'foo', node = '${plus_10:1}'
inter_type = 'plus_10', inter_args = (1,), inter_args_str = ('1',)

    def _evaluate_custom_resolver(
        self,
        key: Any,
        node: Node,
        inter_type: str,
        inter_args: Tuple[Any, ...],
        inter_args_str: Tuple[str, ...],
    ) -> Any:
        from omegaconf import OmegaConf

        resolver = OmegaConf._get_resolver(inter_type)
        if resolver is not None:
            root_node = self._get_root()
            return resolver(
                root_node,
                self,
                node,
                inter_args,
                inter_args_str,
            )
        else:
>           raise UnsupportedInterpolationType(
                f"Unsupported interpolation type {inter_type}"
            )
E           omegaconf.errors.UnsupportedInterpolationType: Unsupported interpolation type plus_10

/Users/dave/micromamba/envs/dvc/lib/python3.11/site-packages/omegaconf/base.py:702: UnsupportedInterpolationType
----------------------------------- Captured log call -----------------------------------
DEBUG    hydra.core.utils:utils.py:279 Setting JobRuntime:name=app
================================ short test summary info ================================
SKIPPED [4] tests/func/utils/test_hydra.py:102: TOML dumper breaks when overriding a list/dict with other type or when handling `null` values.
FAILED tests/func/utils/test_hydra.py::test_compose_and_dump_plugins - omegaconf.errors.UnsupportedInterpolationType: Unsupported interpolation type plus_10
======================== 1 failed, 59 passed, 4 skipped in 0.81s ========================
```

File: dvc/utils/hydra.py [hunk 1]
```diff
@@ -30,16 +31,24 @@ def compose_and_dump(
...
+
     from hydra import compose, initialize_config_dir, initialize_config_module
+    from hydra.core.plugins import Plugins
     from omegaconf import OmegaConf
 
     from .serialize import DUMPERS
 
+    sys.path.append(plugins_path)
+    Plugins.instance()
+    sys.path.pop()
```
Comment (dberenbaum) [inline dvc/utils/hydra.py:50]: This is the only logical change in this PR. It appends the path, loads Hydra plugins, and then resets the path.

Comment (efiop) [inline dvc/utils/hydra.py:50]:
```suggestion
    sys.path.append(plugins_path)
    try:
        Plugins.instance()
    finally:
        sys.path.pop()
```

**Comment by skshetry** [comment]
@dberenbaum, does this introduce any potential security issues for us?

**Comment by efiop** [comment]
@skshetry You tell us 😄

**Comment by dberenbaum** [comment]
@skshetry See the discussion in https://github.com/facebookresearch/hydra/issues/2835. It seems this is left to the user if they wish to enable Hydra plugins. The user must choose to enable Hydra in DVC for this code to run. Then, it is up to them what code they want to import from the Hydra plugins path. If we want to be extra cautious, we could disable this unless the `hydra.plugins_path` is explicitly set.

File: dvc/utils/hydra.py [hunk 2]
```diff
@@ -30,16 +31,26 @@ def compose_and_dump(
...
     from omegaconf import OmegaConf
 
     from .serialize import DUMPERS
 
+    sys.path.append(plugins_path)
+    try:
+        Plugins.instance()
+    finally:
+        sys.path.pop()
+
```
Comment (skshetry) [inline dvc/utils/hydra.py:53]: Can we move this to a separate function, and may be call `load_hydra_plugins`? Or, maybe `load_hydra_plugins_dangerously` to make it more obvious?

**Comment by dberenbaum** [comment]
> Could this be a runtime setting instead of a config? Eg::
> 
> ```
> dvc exp run --hydra-plugins=<path> ...
> ```

I don't think that would be ideal since it would mean having to include that flag every time you run an experiment. What is the benefit?

**Comment by dberenbaum** [comment]
I don't think we are exposing any security risks beyond what Hydra is already doing. The same security risks are present if they run without DVC. If Hydra is enabled, I don't see much reason not to load the plugins since this is the expected behavior of Hydra.

**Comment by skshetry** [comment]
> I don't think we are exposing any security risks beyond what Hydra is already doing. The same security risks are present if they run without DVC. If Hydra is enabled, I don't see much reason not to load the plugins since this is the expected behavior of Hydra.

Reading the issue in the hydra gave me the impression that they don't load custom resolvers by default (because of the same security reason?).

**Review by skshetry — APPROVED** [review]

**Pull request merged**